### PR TITLE
Fix boost on mac and linux: correctly search for boost and set -mt suffix when needed.

### LINKED ISSUE #2823: meson is failing to detect boost libraries suffixed with '-mt' on mac
On macports boost library names look like '/opt/local/lib/libboost_program_options-mt.dylib'.  Boost builds are finding boost, but then saying that the modules are not found.

### PR DESCRIPTION
**Pull request by bredelings**
Currently meson incorrect determines that the modname for boost program options is boost_program_options-mt instead of boost_program_options if the filename is `libboost_program_options-mt.dylib`.  This makes meson unable to detect boost installed from macports.

Fix this by correctly computing the modname from the filename, and factor out all the code that does this into one place instead of doing in a different way in each place.

Fixes #2823

### CODE REVIEW

File: mesonbuild/dependencies/misc.py [hunk 1]
```diff
@@ -296,22 +303,11 @@ def detect_lib_modules_win(self):
...
-            (_, fname) = os.path.split(entry)
-            modname = fname.split('-', 1)
-            if len(modname) > 1:
-                modname = modname[0]
-            else:
-                modname = modname.split('.', 1)[0]
-            if self.static:
-                modname = modname[3:]
-            self.lib_modules[modname] = fname
+            self.lib_modules[self.modname_from_filename(fname)] = fname
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

[F821] undefined name 'fname'

File: mesonbuild/dependencies/misc.py [hunk 2]
```diff
@@ -296,22 +303,11 @@ def detect_lib_modules_win(self):
...
         if len(globber2_matches) == 0:
             for entry in glob.glob(os.path.join(self.libdir, globber1 + '.lib')):
-                (_, fname) = os.path.split(entry)
-                modname = fname.split('.', 1)[0]
                 if self.static:
-                    modname = modname[3:]
-                    self.lib_modules[modname] = fname
+                    self.lib_modules[self.modname_from_filename(fname)] = fname
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

[F821] undefined name 'fname'

**Comment by bredelings** [comment]
BTW, I replaced `(,fname) = os.path.split(entry)` with `fname = os.path.basename(entry)` since the latter seems more readable.  Is there some reason to prefer the former?

**Comment by bredelings** [comment]
Fixes #2823

**Comment by nirbheek** [comment]
@NickeZ please review :)

**Comment by mojca** [comment]
@bredelings: I applied this patch on top of 0.44.0 and then tried to build [bali-phy](https://github.com/bredelings/Bali-Phy), but I'm still getting the "not found" error when using `-Dwith-system-boost=True`.

```
The Meson build system
Version: 0.44.0
Source dir: ...
Build dir: ...
Build type: native build
Project name: bali-phy
Native C compiler: /usr/bin/clang (clang 9.0.0)
Appending CFLAGS from environment: '-pipe -Os -arch x86_64'
Appending LDFLAGS from environment: '-L/opt/local/lib -Wl,-headerpad_max_install_names -arch x86_64'
Appending CPPFLAGS from environment: '-I/opt/local/include'
Native C++ compiler: /usr/bin/clang++ (clang 9.0.0)
Appending CXXFLAGS from environment: '-pipe -Os -stdlib=libc++ -arch x86_64'
Appending LDFLAGS from environment: '-L/opt/local/lib -Wl,-headerpad_max_install_names -arch x86_64'
Appending CPPFLAGS from environment: '-I/opt/local/include'
Build machine cpu family: x86_64
Build machine cpu: x86_64
Has header "sys/resource.h": YES
Has header "fenv.h": YES
Checking for function "log2": YES
Configuring config.h using configuration
Found pkg-config: /opt/local/bin/pkg-config (0.29.2)
Native dependency cairo found: YES 1.14.12

Meson encountered an error in file meson.build, line 76, column 2:
Requested Boost library 'program_options' not found
```

**Comment by NickeZ** [comment]
Overall looks good to me. As long as the tests pass I would merge it.

It is unfortunate that we don't have better tests since it is difficult to see if this changes the behavior in all the edge cases. Like for example, if I have `program_options`, `program_options-mt` and `program_options-mt-gd` installed. Does it still pick the correct library if dependency is called with `static=true` and/or `threaded=multi`.

**Comment by bredelings** [comment]
The not-found error is a regression in meson handling of -mt, so 
interesting that you should mention that.  We could probably merge this 
after meson 0.44.1 is released, since I think it will contain the patch 
from here (https://github.com/mesonbuild/meson/issues/2823) and will (I 
suspect) be released in a week or so.

-BenRI

**Comment by bredelings** [comment]
We do have `test cases/frameworks/1 boost`.  Why did this not trigger in travis?

**Comment by bredelings** [comment]
@mojca I'll check again later, but for now did you set BOOST_ROOT=/opt/local ?

**Comment by bredelings** [comment]
For me, this does not work:
`CPPFLAGS=-I/opt/local/include LDFLAGS=-L/opt/local/lib meson ../master/ meson -Dwith-system-boost=True`
But this does work:
`BOOST_ROOT=/opt/local CPPFLAGS=-I/opt/local/include LDFLAGS=-L/opt/local/lib meson ../master/ meson -Dwith-system-boost=True`
It seems wrong to have to set an environment variable specifying the root for every library - is there some way to set this globally, like for cross builds?

Hmm... it seems that the way searching for boost works, it picks some specific dirs and checks if they contain files with names that match a pattern.  So, LDFLAGS has no effect on this search.

**Comment by mojca** [comment]
Indeed, with `BOOST_ROOT=/opt/local` it (kind of) works:
```
Dependency Boost (program_options, random, system, filesystem, chrono) found: YES 1.65.1, /opt/local
```
That is: the configure goes through, but the linker later fails, being unable to find the library because the `-mt` suffix is missing:
```
$ /usr/bin/clang++  -o src/alignment-info 'src/alignment-info@exe/tools_alignment-info.cc.o' -L/opt/local/lib -pipe -Os -stdlib=libc++ -arch x86_64 -Wl,-headerpad_max_install_names -arch x86_64 src/libbali-phy.a -L/opt/local/lib -lboost_program_options -lboost_random -lboost_system -lboost_filesystem -lboost_chrono -L/opt/local/lib -lboost_program_options -lboost_random -lboost_system -lboost_filesystem -lboost_chrono
$ ld: library not found for -lboost_program_options
```
And I totally agree that having to manually specify the prefix for every single dependency feels super weird to me. Sure, it makes sense to specify it when I want to take a dependency from elsewhere, but from the perspective of a package manager where one knows in advance that all dependencies should come from one certain well-known prefix, it feels waste of time and very error-prone to have to specify it for every single dependency.

**Comment by NickeZ** [comment]
@mojca Which OS are you using that puts boost in /opt/local? That sounds like an unusual place for a standard directory.

@bredelings Yes, I'm aware of the tests we have. But they only test the very simple case where boost is installed with the system provided package manager (linux) or the binaries provided from SF.net (windows). There are many other ways you can build and install boost which we don't test.

**Comment by mojca** [comment]
I'm using macOS with the [MacPorts](https://github.com/macports/) package manager. The path `/opt/local` is default (but it could be any given path and I particularly don't like the path being hardcoded anywhere). The reason why the path might sound a tiny bit strange is that all the "usual" prefixes are reserved for binaries provided by Apple.

I believe it should be straightforward to write a test for this scenario. Just install a super simple hello world library with that name (`libboost_program_options-mt.dylib`) under any given prefix (it could be in `/tmp` or under the directory where the tests are being run) and try to find the library and link against it. No need to actually test a fully-functional boost installation. (Of course just as one of many possible tests.)

**Comment by bredelings** [comment]
@mojca Would the link work if you _also_ supply LDFLAGS specifically?  The line I showed about (with CXXFLAGS and LDFLAGS) linked for me.

**Comment by bredelings** [comment]
@NickeZ Sorry if I mis-spoke.  I'm just confused due to inexperience - why aren't the travis tests currently failing on osx?  Is it the case that travis doesn't run tests under `test cases`?

**Comment by mojca** [comment]
@bredelings: The `LDFLAGS` have been set as can also be seen from the command being executed. The problem is in the
```
-lboost_program_options
```
flag which should have been
```
-lboost_program_options-mt
```
instead.

What's the output of `otool -L /opt/local/bin/bali-phy` on your machine?

**Comment by NickeZ** [comment]
I wasn't aware that the tests we have weren't running on mac. I'm looking into it.

**Comment by bredelings** [comment]
@mojca Good catch.  On my system, we detect the '-mt' form, but link against the non-mt form, and both exist. Let me look into this...

**Comment by bredelings** [comment]
@NickeZ I think we might run the unit-tests, but not the `test cases` tests?

**Comment by bredelings** [comment]
Argh.  The code in detect_modules can be consistent with either validate_requested or get_link_args( ), but not both.  Also, the get_link_args( ) routine calling compiler.find_library('boost_'+module_name).  If we are going to call compiler.find_library( ) ignore, we should call it in detect_modules( ). 

One other issues is that almost any change to the meson.build file results in 30 seconds of waiting, if we have a boost dependency.  Maybe this is because we are calling compiler.find_library( ) from get_link_args( )?

**Comment by mojca** [comment]
[They are](https://travis-ci.org/mesonbuild/meson/jobs/320460631).

File: mesonbuild/dependencies/misc.py [hunk 3]
```diff
@@ -334,19 +332,23 @@ def detect_lib_modules_nix(self):
...
                 elif self.is_multithreading and entry.endswith('-mt.{}'.format(libsuffix)):
-                    self.lib_modules[name] = lib
+                    pass
                 elif not entry.endswith('-mt.{}'.format(libsuffix)):
-                    self.lib_modules[name] = lib
+                    pass
+                else:
+                    continue
+                modname = self.modname_from_filename(entry)
+                if not modname in self.lib_modules:
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

[E713] test for membership should be 'not in'

**Comment by bredelings** [comment]
@mojca @NickeZ The call to compiler.find_library( ) in get_link_args( ) was throwing away the '-mt' suffix, so I switched to just using the files we had already discovered in detect_lib_modules_nix( ).  This incidentally is much faster, because compiler.file_library( ) is very slow and get_link_args( ) is called a huge number of times.

**Comment by jpakkane** [comment]
Tests with dependencies are not run on Mac or Windows by default because installing all the dependencies reliably is a pain and the build machines are stateless.

**Comment by bredelings** [comment]
The travis build shows no errors except a timeone in one of the 6 jobs - can someone restart this?

**Comment by bredelings** [comment]
@nirbheek @NickeZ I think a number of further improvements are possible.  These would include fixing cross builds, finding boost if its on in /usr/lib or /usr/local/lib, etc.
* Should these be discussed on the mailing list, or by opening new issues (or both)?
* Should I just push changes to this PR, or open new PRs?

**Comment by NickeZ** [comment]
@bredelings You can probably improve the situation as much as possible for mac in this PR. But if you plan to implement cross-compilation I think that should go into a separate PR. This is not really my call anyway since I'm only a contributer.

@jpakkane Boost is installed with homebrew on the jenkins macos build images. Should we start testing against it?

**Comment by jpakkane** [comment]
If the way the CI image ships it is representative with how people are using it in the real world, why not.

**Comment by ignatenkobrain** [comment]
Flake8 detected 6 issues on efb45b251da4bce2f1d2ebd357356c768d3e0608
Visit https://sideci.com/gh/19784232/pull_requests/2824 to review the issues.

**Comment by bredelings** [comment]
After the last change we can find boost in the standard location on macports without needing to set BOOST_ROOT=/opt/local

**Comment by bredelings** [comment]
@NickeZ Enabling checks on homebrew is a good idea.  I have remote access to a Mac, though it is not my own, so perhaps I can help.

File: mesonbuild/dependencies/misc.py [hunk 4]
```diff
@@ -25,7 +25,7 @@
 
 from .. import mlog
 from .. import mesonlib
-from ..mesonlib import Popen_safe, extract_as_list
+from ..mesonlib import Popen_safe, extract_as_list, EnvironmentException
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

[F401] '..mesonlib.Popen_safe' imported but unused

File: mesonbuild/dependencies/misc.py [hunk 5]
```diff
@@ -231,25 +264,30 @@ def validate_requested(self):
 
     def detect_version(self):
         try:
-            ifile = open(os.path.join(self.incdir, 'boost', 'version.hpp'))
-        except FileNotFoundError:
+            version = self.compiler.get_define('BOOST_LIB_VERSION', '#include <boost/version.hpp>', self.env, self.get_compile_args(), [])
+        except EnvironmentException:
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

[F821] undefined name 'EnvironmentException'

**Comment by bredelings** [comment]
@ignatenkobrain I'm getting an error from ci/sideci because I am importing an exception that I only catch, but do not throw.  If I don't import it, I get another error for not defining it.  What should I do?

File: mesonbuild/dependencies/misc.py [hunk 6]
```diff
@@ -172,10 +213,8 @@ def detect_win_roots(self):
         return res
 
     def detect_nix_incdir(self):
-        for root in self.boost_roots:
-            incdir = os.path.join(root, 'include', 'boost')
-            if os.path.isdir(incdir):
-                return os.path.join(root, 'include')
+        if self.boost_root:
+            return os.path.join(self.boost_root,'include')
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

[E231] missing whitespace after ','

File: mesonbuild/dependencies/misc.py [hunk 7]
```diff
@@ -958,3 +1013,117 @@ def get_methods(self):
     'boost_type_erasure',
     'boost_wave'
 ]
+
+BOOST_DIRS = [
+#    'boost',
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

[E122] continuation line missing indentation or outdented

**Comment by bredelings** [comment]
@mojca @NickeZ With these changes, I can build for either macports or homebrew with them both installed.  I'd like to get the boost test for homebrew working.  This change to `run_project_tests.py` enables the test:
```
-    elif mesonlib.is_osx() or mesonlib.is_cygwin():
+    elif mesonlib.is_osx():
+        # Just do the BOOST test
+        gathered_tests += [('framework', ['test cases/frameworks/1 boost'], False)]
+    elif mesonlib.is_cygwin():
+        # Skip all the framework tests
```
However, the test fails, saying that `#include <boost/version.hpp>` can't find the file.  On the mac I have access to, going into the boost dir and typing 'meson build' works though.

Any ideas?  I thought that homebrew boost was installed already on travis osx images (based on @NickeZ's boost-brew branch), so I would expect `/usr/local/include/boost/version.hpp` to exist.  I hear that on recent osx versions, you have to execute `xcode-select --install` in order to make the cli tools search /usr/local, so maybe we have to do this first?

**Comment by bredelings** [comment]
@jpakkane Now that we can test boost on mac, is this OK?  If there is anything you'd like me to change about the approach please let me know.

**Comment by jpakkane** [comment]
Looks good, but I'm not a fan of hardcoding the list of directories because it will get out of date. Could we instead get it from the system when running the checkup the first time? Something like:

    BOOST_DIRS=None
    ...
    if BOOST_DIRS is None:
        BOOST_DIRS = <glob from filesystem>
    ...
    <rest of code here>

**Comment by bredelings** [comment]
@jpakkane My rationale is
* Since the warning is just to avoid failing on old `meson.build` files that list non-library modules, I don't think we need to update the list of boost dirs.  The code above says 'This will be an error in the future', so it seems like a deprecation warning.  Presumably after a year we can delete the BOOST_DIRS and the warning as well?
* On Unix we don't always know which directory contains the header files, so I don't know how to find the list of dirs from the filesystem in that case.  (If we could somehow get the list of directories that the compiler searches when we provide a list of -Idir flags, then we could probably search these dirs ourselves, but I don't know how to do this.)
Does that seem sensible?

This raises another problem that could perhaps be dealt with in the future.  I don't think we can in general give an error if people give a module that we don't know, because it could be a library from a future version of boost.  The library also might not be a directory name under /usr/include/boost/ even in the future version of boost -- consider `libboost_test_exec_monitor.a`.  What we can do is issue a warning that we don't recognize the module as a library, assume the module is a library anyway, and fail if we can't link against the library.  Thoughts?

**Comment by jpakkane** [comment]
Let's merge this. We can fix issues later if they arise. Thanks.
